examples/directx12_example/imgui_impl_dx12.cpp

```diff
@@ -0,0 +1,472 @@
+}
+
+void ImGui_ImplDX12_SetRenderData(ID3D12GraphicsCommandList* _commandList, D3D12_CPU_DESCRIPTOR_HANDLE _renderTarget)
+{
+    g_commandList = _commandList;
+    g_renderTarget = _renderTarget;
+}
+
+void ImGui_ImplDX12_RenderDrawLists(ImDrawData* _draw_data)
+{
+    // Range CPU will read from mapping the upload buffer
+    // End < Begin specifies CPU will not read the mapped buffer
+    D3D12_RANGE readRange;
+    readRange.End = 0;
+    readRange.Begin = 1;
+
+    char* mappedBuffer = 0;
+    g_uploadBuffer->Map(0, &readRange, (void**)&mappedBuffer);
+    if (mappedBuffer == NULL)
+        return;
+
+    char* writeCursor = mappedBuffer;
+
+    // Copy the projection matrix at the beginning of the buffer
+    {
+        float translate = -0.5f * 2.f;
+        const float L = 0.f;
+        const float R = ImGui::GetIO().DisplaySize.x;
+        const float B = ImGui::GetIO().DisplaySize.y;
+        const float T = 0.f;
+        const float mvp[4][4] = 
+        {
+            { 2.0f/(R-L),   0.0f,           0.0f,       0.0f},
+            { 0.0f,         2.0f/(T-B),     0.0f,       0.0f,},
+            { 0.0f,         0.0f,           0.5f,       0.0f },
+            { (R+L)/(L-R),  (T+B)/(B-T),    0.5f,       1.0f },
+        };
+
+        memcpy(writeCursor, &mvp[0], sizeof(mvp));
+        writeCursor += sizeof(mvp);
+    }
+
+    // Copy the vertices and indices for each command list
+    for (int n = 0; n < _draw_data->CmdListsCount; n++)
+    {
+        const ImDrawList* cmd_list = _draw_data->CmdLists[n];
+        size_t verticesCount = cmd_list->VtxBuffer.size();
+        size_t indicesCount  = cmd_list->IdxBuffer.size();
+        size_t verticesSize = verticesCount * sizeof(ImDrawVert);
+        size_t indicesSize = indicesCount * sizeof(ImDrawIdx);
+
+        // Copy the vertex data
+        memcpy(writeCursor, &cmd_list->VtxBuffer[0], verticesSize);
+        writeCursor += verticesSize;
+
+        // Copy the index data
+        memcpy(writeCursor, &cmd_list->IdxBuffer[0], indicesSize);
+        writeCursor += indicesSize;
+    }
+
+    ID3D12GraphicsCommandList* commandList = g_commandList;
+
+    D3D12_VIEWPORT viewport;
+    viewport.Width = ImGui::GetIO().DisplaySize.x;
+    viewport.Height = ImGui::GetIO().DisplaySize.y;
+    viewport.MinDepth = 0.f;
+    viewport.MaxDepth = 1.f;
+    viewport.TopLeftX = 0.f;
+    viewport.TopLeftY = 0.f;
+
+    commandList->RSSetViewports(1, &viewport);
+    commandList->OMSetRenderTargets(1, &g_renderTarget, FALSE, nullptr);
+    commandList->IASetPrimitiveTopology(D3D_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
+    commandList->SetPipelineState(g_pipelineState);
+
+    D3D12_GPU_VIRTUAL_ADDRESS bufferAddress = g_uploadBuffer->GetGPUVirtualAddress();
+    commandList->SetGraphicsRootConstantBufferView(1, bufferAddress);
+
+    uint64_t readCursor = 64; // Our constant buffer takes 64 bytes - one mat4x4
+
+    for (int n = 0; n < _draw_data->CmdListsCount; n++)
```

Comment:
This for loop can be merged with the previous one. So you copy the render data in parallel with filling the command buffer.

also you never unmapped `g_uploadBuffer`